docs_nnx/guides/randomness.md

```diff
@@ -0,0 +1,245 @@
+
+nnx.display(params_key_state)
+```
+
+## Reseeding
+In Haiku and Flax Linen, random state explicitly passed to `apply` each time before calling the model. This makes it easy to control the randomness of the model when needed e.g. for reproducibility. In Flax NNX there are two ways to approach this:
+1. By passing an `Rngs` object through the `__call__` stack manually. Standard layers like `Dropout` and `MultiHeadAttention` accept an `rngs` argument in case you want to have tight control over the random state.
+2. By using `nnx.reseed` to set the random state of the model to a specific configuration. This option is less intrusive and can be used even if the model is not designed to enable manual control over the random state.
+
+`reseed` is a function that accepts an arbitrary graph node (this include pytrees of Flax NNX Modules) and some keyword arguments containing the new seed or key value for the `RngStream`s specified by the argument names. `reseed` will then traverse the graph and update the random state of the matching `RngStream`s, this includes both setting the `key` to a possibly new value and resetting the `count` to zero.
+
+Here's an example of how to using `reseed` to reset the random state of the `Dropout` layer and verify that the computation is identical to the first time the model was called:
+
+```{code-cell} ipython3
+model = Model(nnx.Rngs(params=0, dropout=1))
+x = jnp.ones((1, 20))
+
+y1 = model(x)
+y2 = model(x)
+
+nnx.reseed(model, dropout=1) # reset dropout RngState
+y3 = model(x)
+
+assert not jnp.allclose(y1, y2) # different
+assert jnp.allclose(y1, y3)     # same
+```
+
+## Splitting Rngs
+When interacting with transforms like `vmap` or `pmap` it is often necessary to split the random state such that each replica has its own unique state. This can be done in two way, either by manually splitting a key before passing it to one of the `Rngs` streams as shown below:
+
+```{code-cell} ipython3
+split_key = random.split(random.key(0), 5)
+rngs = nnx.Rngs(params=0, dropout=split_key)
+nnx.display(rngs)
+```
+
+Notice that both the `key` and the `count` for the `dropout` stream have a dimension of `5`. This strategy the benefit of being very explicit, but can be a bit limiting as it wouldn't allow the stream to be used outside of the transform:
+
+```{code-cell} ipython3
+try:
+  rngs.dropout()
+except ValueError as e:
+  print(e)
+```
+
+The other option is to use the `nnx.split_rngs` decorator which will automatically split the random state of any `RngStream`s found in the inputs of the function, and will automatically "lower" them once the function call ends so the `Rngs` can be used outside again. `split_rngs` allows passing Filters to the `only` keyword argument to select the `RngStream`s that should be split. Using `split_rngs` is useful in combination with a transform but here we will show a simple example without any transforms to illustrate the concept, we'll use `split_rngs` with a transform on the next section.
```

Comment:
I have a feeling that we should introduce and suggest `nnx.split_rngs` first, since it works better with transform cases and doesn't have the drawback that splitting in stream has (aka. can't work outside transform).

Actually, if there isn't any case in which splitting in stream is better, maybe it's fine to simply avoid introducing that... or introducing that as a lower-level view of what `nnx.split_rngs` actually does.

Reply:
Good point. I'll think about it. We might as well just delete the other example as stick with `split_rngs`.
The main motivation of showing how to manually lift was to make it more familiar as the operation is explicit but it might be less used.